test/suite/test_timestamp22.py

```diff
@@ -0,0 +1,444 @@
+
+        # If the final commit is too old, we'll fail.
+        if commit_ts < self.oldest_ts or commit_ts < self.stable_ts:
+            ok_commit = False
+
+        # ODDITY: We don't have to move the commit_ts ahead, but it has to be
+        # at least the value of the first commit timestamp set.
+        if commit_ts < first_commit_ts:
+            ok_commit = False
+
+        # If a prepare fails or any setting of timestamp fails, the commit fails as well.
+        if not ok_prepare or not ok_tstxn1 or not ok_tstxn2:
+            ok_commit = False
+
+        msg = 'inserts with commit config(' + commit_config + ')'
+
+        try:
+            for i in range(0, self.nrows):
+                needs_rollback = False
+                if self.do_illegal():
+                    # Illegal outside of transaction
+                    self.report('prepare_transaction', prepare_config)
+                    with self.expect(False, 'prepare outside of transaction'):
+                        session.prepare_transaction(prepare_config)
+
+                with self.expect(True, 'begin_transaction(' + begin_config + ')'):
+                    session.begin_transaction()
+                    needs_rollback = True
+
+                if do_tstxn1:
+                    with self.expect(ok_tstxn1, 'timestamp_transaction(' + tstxn1_config + ')'):
+                        session.timestamp_transaction(tstxn1_config)
+
+                self.report('set key/value')
+                with self.expect(True, 'cursor insert'):
+                    cursor[ds.key(i)] = value
+
+                if do_tstxn2:
+                    with self.expect(ok_tstxn2, 'timestamp_transaction(' + tstxn2_config + ')'):
+                        session.timestamp_transaction(tstxn2_config)
+
+                if do_prepare:
+                    self.report('prepare_transaction', prepare_config)
+                    with self.expect(ok_prepare, 'prepare'):
+                        session.prepare_transaction(prepare_config)
+
+                # Doing anything else after the prepare, like a timestamp_transaction(), will fail
+                # with a WT panic.  Don't do that, or else we can't do anything more in this test.
+
+                # If we did a successful prepare and are set up (by virtue of bad timestamps)
+                # to do a bad commit, WT will panic, and the test cannot continue.
+                # Only proceed with the commit if we have don't have that particular case.
+                if ok_commit or not do_prepare or not ok_prepare:
+                    needs_rollback = False
+                    self.report('commit_transaction', commit_config)
+                    with self.expect(ok_commit, 'commit'):
+                        session.commit_transaction(commit_config)
+                        self.commit_value = value
+                if needs_rollback:
+                    self.report('rollback_transaction')
+                    needs_rollback = False
```

Comment:
The multiple uses of `if needs_rollback:` so close together deserve a comment either here or later on. It is a complex depth of potential code paths.

Reply:
Thanks for catching this.  Turns out I don't need the extra call to do rollback outside of the loop.